order/basic.lean

```diff
@@ -26,8 +26,65 @@ theorem monotone_comp {f : α → β} {g : β → γ} (m_f : monotone f) (m_g :
   monotone (g ∘ f) :=
 assume a b h, m_g (m_f h)
 
+lemma monotone_mul_nonneg {R : Type*} [decidable_linear_ordered_comm_ring R] {a : R} (a_nonneg : 0 ≤ a) : monotone (λ x, a*x) :=
+assume b c b_le_c, mul_le_mul_of_nonneg_left b_le_c a_nonneg
 end monotone
 
+section max
+variable [decidable_linear_order α] 
+
+lemma le_max_of_le_left {a b : α} ( c: α ) (H : a ≤ b ) : a ≤ max b c :=
+le_trans H (le_max_left b c)
+
+lemma le_max_of_le_right {a c : α} (b :α) (H : a ≤ c) : a ≤ max b c :=
+le_trans H (le_max_right b c)
+
+lemma max_le_max  {a b c d  : α} (h1 : a ≤ b) (h2 : c ≤ d) : max a c ≤ max b d := 
+max_le (le_trans h1 (le_max_left b d)) (le_trans h2 (le_max_right b d))
+
+lemma max_monotone_fun [decidable_linear_order β] {f : α → β} (H : monotone f) (a a' : α) :
+max (f a) (f a') =  f(max a a') :=
+begin
+by_cases a ≤ a',
+{ have fa_le_fa' := H h,
+  rw max_comm,
+  rw max_eq_left fa_le_fa',
+  have T :=  max_eq_left h,
+  rw max_comm at T,
+  rw T },
+{ have h' : a' ≤ a := le_of_not_ge h,
+  rw max_eq_left (H h'),
+  rw  max_eq_left h' }
+end
+
+variables {R : Type*} [decidable_linear_ordered_comm_ring R]
+
+lemma max_mul_nonneg {a : R} (b c : R) (a_nonneg : 0 ≤ a) : max (a*b) (a*c) = a*(max b c) :=
+max_monotone_fun (monotone_mul_nonneg a_nonneg) b c
+
+lemma max_mul_le_mul_max {a b c d  : R} (ha : 0 ≤ a) (hb : 0 ≤ b) (hd: 0 ≤ d) : 
+max (a*b) (c*d) ≤ (max a c) * (max b d) :=
```

Comment:
The assumption hb is not necessary here.

Reply:
Thanks Sébastien. In my use case all four are nonnegative and I didn't push optimization all the way through. Now it's fixed and I even took the opportunity to write a proof Mario and Johannes may find delightfully obfuscated.